near-sdk/src/promise.rs

```diff
@@ -365,9 +398,65 @@ impl Promise {
     }
 
     #[cfg(feature = "global-contracts")]
+    /// Publish contract code to the global contract registry, indexed by hash only.
+    ///
+    /// This is a high-level API that publishes contract code to make it globally accessible.
+    /// The code can then be referenced by hash using [`Self::deploy_from_published`],
+    /// saving on deployment costs when deploying multiple instances.
+    ///
+    /// Uses low-level [`crate::env::promise_batch_action_deploy_global_contract`]
+    ///
+    /// # Examples
+    /// ```no_run
+    /// use near_sdk::{Promise, NearToken};
+    ///
+    /// let code = vec![0u8; 100]; // Contract bytecode
+    ///
+    /// // Publish contract code - can be referenced by its hash later
+    /// Promise::new("alice.near".parse().unwrap())
+    ///     .create_account()
+    ///     .publish_contract_by_hash(code)
+    ///     .detach();
+    /// ```
+    pub fn publish_contract_by_hash(self, code: Vec<u8>) -> Self {
+        self.add_action(PromiseAction::DeployGlobalContract { code })
+    }
+
+    #[cfg(feature = "global-contracts")]
+    /// Publish contract code to the global contract registry, indexed by the current account.
+    ///
+    /// This is a high-level API that publishes contract code and associates it with the
+    /// current account. The code can then be referenced by account ID using
+    /// [`Self::deploy_from_published`], making it easier for others to find and use.
+    ///
+    /// Uses low-level [`crate::env::promise_batch_action_deploy_global_contract_by_account_id`]
+    ///
+    /// # Examples
+    /// ```no_run
+    /// use near_sdk::{Promise, NearToken};
+    ///
+    /// let code = vec![0u8; 100]; // Contract bytecode
+    ///
+    /// // Publish and associate with account - others can reference by account ID
+    /// Promise::new("multisig-template.near".parse().unwrap())
+    ///     .create_account()
+    ///     .publish_contract_by_account(code)
```

Comment:
I tried to look at this example without any prior context of global contracts and I find it impossible to understand. Let's brainstorm it more in a chat.